xarray/backends/api.py

```diff
@@ -352,11 +344,25 @@ def maybe_decode_store(store, lock=False):
         with close_on_error(store):
             ds = maybe_decode_store(store)
     else:
-        if engine is not None and engine != 'scipy':
-            raise ValueError('can only read file-like objects with '
-                             "default engine or engine='scipy'")
-        # assume filename_or_obj is a file-like object
-        store = backends.ScipyDataStore(filename_or_obj)
+        if engine not in [None, 'scipy', 'h5netcdf']:
+            raise ValueError('can only read bytes or file-like objects with '
+                             "engine = None, 'scipy', or 'h5netcdf'")
+        else:
+            if isinstance(filename_or_obj, bytes):
+                filename_or_obj = BytesIO(filename_or_obj)
+            # read first bytes of file-like object to determine engine
```

Comment:
It would be nice to factor this logic (for file-type detection) out into a separate helper function, just to keep `open_dataset()` a little tidier.

Reply:
i've done this now and added some new error messages and tests. one of the things i did is not cast to BytesIO() behind the scenes, because this was leading to confusing errors if open_dataset() was called more than once and the file/buffer was then no longer at the initial read position. i guess the alternative is adding a line that forces `filename_or_obj.seek(0)`...